capi-configuration/README.md

```diff
@@ -0,0 +1,333 @@
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: machinepool-21-0-0
+  namespace: giantswarm
+data:
+  spec:
+    replicas: 3
+    template:
+      spec:
+        version: v1.21.4
+```
+It is important to note that a assumption is that **no** values in the base file are immutable!
+
+### Cluster-api context dependent values
+
+We have some cases in `cluster-api` where a value in a custom resource is neither part of the fundamental structure nor is it easy to statically default.
+
+These values are usually highly dependent on the context in which the cluster is created in.
+For example the `CIDR` in which the cluster is created.
+We currently assume that these values will be defaulted by custom webhooks or operators with significant provider specific domain knowledge.
+
+### Putting together a Giant Swarm release
+
+A Giant Swarm releases is just a set of `cluster-api` base files as defined previously with added `apps`.
+
+The following folder structure would be contained in `giantswarm/example-customer`:
+```yaml
+aws:
+  v21.0.0:
+    base:
+      cluster.yaml # configmap containing the base file (defined above)
+      aws-cluster.yaml
+      machinepool.yaml
+      awsmachinepool.yaml
+      ...
+      net-exporter.yaml # configmap containing version information for defaulting!
+    overlay:
+      special-overlay.yaml # Added by customer/AE/SA to e.g. overlay a different flatcar version!
+  v22.0.0:
+    base:
+      ...
+    overlay:
+      ...
+```
+
+In the management cluster a release is then simply a set of configmaps.
+The configmaps for e.g. release `v21.0.0` will be applied through argoCD by taking the `base` configmaps, applying any `overlay` from a customer and then deploying them.
+
+As shown in previous examples these configmaps would contain the release in their name (e.g. `machinepool-21-0-0`).
+Any in-cluster webhook or operator can then easily navigate to these configmaps to look up defaults for a cluster.
+
+A cluster creation workflow can now look as follows:
+1. Client (e.g. `happa`) submits `--dry-run` with `fundamental` custom resources to management cluster.
+2. Mutating webhooks take information from `base file` configmaps to default.
+3. Client receives fully defaulted custom resources and allows users to make manual changes.
+4. Client submits custom resources
+5. Cluster is created.
+
+A cluster upgrade workflow can then look like this:
+1. Client changes cluster version label.
+2. `upgrade-operator` reads `base file` configmaps according to the new version.
+3. `upgrade-operator` starts replacing fields with new configuration from configmaps.
+
+## Going gitops all the way
+
+A next step in `gitops` adoption would be the addition of having the cluster custom resources in `gitops` as well.
+So the shared git repository would contain not only release information but also the clusters themselves.
+
+```yaml
+gauss: # installation
+  peu01: #clusterID
+    base: # managed fully by GS
```

Comment:
Same question here -- why do we want to duplicate a static base between customers?